apps/trivium/src/trivium/trivium_byte.rs

```diff
@@ -0,0 +1,258 @@
+
+        // Initialization of Trivium registers: a has the secret key, b the input vector,
+        // and c a few ones.
+        let mut a_byte_reg = [0u8; 12].map(|x| FheUint8::encrypt_trivial(x));
+        let mut b_byte_reg = [0u8; 11].map(|x| FheUint8::encrypt_trivial(x));
+        let mut c_byte_reg = [0u8; 14].map(|x| FheUint8::encrypt_trivial(x));
+
+        for i in 0..10 {
+            a_byte_reg[12 - 10 + i] = key[i].clone();
+            b_byte_reg[11 - 10 + i] = FheUint8::encrypt_trivial(iv[i]);
+        }
+
+        // Magic number 14, aka 00001110: this represents the 3 ones at the beginning of the c
+        // registers, with additional zeros to make the register's size a multiple of 8.
+        c_byte_reg[0] = FheUint8::encrypt_trivial(14u8);
+
+        unset_server_key();
+        let mut ret = TriviumStreamByte::<FheUint8>::new_from_registers(
+            a_byte_reg,
+            b_byte_reg,
+            c_byte_reg,
+            Some(server_key.clone()),
+        );
+        ret.init();
+        ret
+    }
+}
+
+impl<T> TriviumStreamByte<T>
+where
+    T: TriviumByteInput<T> + Send,
+    for<'a> &'a T: TriviumByteInput<T>,
+{
+    /// Internal generic contructor: arguments are already prepared registers, and an optional FHE
+    /// server key
+    fn new_from_registers(
+        a_register: [T; 12],
+        b_register: [T; 11],
+        c_register: [T; 14],
+        sk: Option<ServerKey>,
+    ) -> Self {
+        Self {
+            a_byte: StaticBitDeque::<12, T>::new(a_register),
+            b_byte: StaticBitDeque::<11, T>::new(b_register),
+            c_byte: StaticBitDeque::<14, T>::new(c_register),
+            fhe_key: sk,
+        }
+    }
+
+    /// The specification of Trivium includes running 1152 (= 18*64) unused steps to mix up the
+    /// registers, before starting the proper stream
+    fn init(&mut self) {
+        for _ in 0..18 {
+            self.next_64();
+        }
+    }
+
+    /// Computes 8 potential future step of Trivium, b*8 terms in the future. This does not update
+    /// registers, but rather returns with the output, the three values that will be used to
+    /// update the registers, when the time is right. This function is meant to be used in
+    /// parallel.
+    fn get_output_and_values(
+        a_byte: StaticBitDeque<12, T>,
+        b_byte: StaticBitDeque<11, T>,
+        c_byte: StaticBitDeque<14, T>,
+        b: usize,
+    ) -> [T; 4] {
+        let n = b * 8 + 7;
+        assert!(n < 65);
+
+        let ((a1, a2, a3, a4, a5), ((b1, b2, b3, b4, b5), (c1, c2, c3, c4, c5))) = rayon::join(
+            || Self::get_bytes(a_byte, [91 - n, 90 - n, 68 - n, 65 - n, 92 - n]),
+            || {
+                rayon::join(
+                    || Self::get_bytes(b_byte, [82 - n, 81 - n, 77 - n, 68 - n, 83 - n]),
+                    || Self::get_bytes(c_byte, [109 - n, 108 - n, 86 - n, 65 - n, 110 - n]),
+                )
+            },
+        );
```

Comment:
[a_byte, b_bytes, c_byte].par_iter().zip([array of indices ordered properly].par_iter()) ?

Not sure if nested joins are discouraged or not, I just find it awkard, unless you benched it and perf looks better with this rather than parallel iterators

general remak for nested joins, I let you check if there is a way to update the nested join pattern to parallel iterators, and if it has some performance implications, hopefully the code is not less readable with parallel iterators, I let you be the judge of that, performance being criterion number 1 to select one pattern over the other

Reply:
I agree it's ugly. I'm looking for ways to remove the nested join pattern as much as possible ! I think par_iter has the same performance and readability, when applicable. What is discourage by the Rayon doc is using `scope` (the ideal API here): they clearly state it's more convenient but less efficient

Reply:
if it makes sense, no need to waste time if it's already ok and there is no real benefit